<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Malachy Campbell" />


<title>Day1_1020.1200 - GWAS exercises</title>

<script src="Day1Ex_Final_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="Day1Ex_Final_files/bootstrap-3.3.6/css/bootstrap.min.css" rel="stylesheet" />
<script src="Day1Ex_Final_files/bootstrap-3.3.6/js/bootstrap.min.js"></script>
<script src="Day1Ex_Final_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<script src="Day1Ex_Final_files/navigation-1.1/tabsets.js"></script>
<script src="Day1Ex_Final_files/navigation-1.1/codefolding.js"></script>
<link href="Day1Ex_Final_files/magnific-popup-1.1.0/magnific-popup.css" rel="stylesheet" />
<script src="Day1Ex_Final_files/magnific-popup-1.1.0/jquery.magnific-popup.min.js"></script>
<link href="Day1Ex_Final_files/clean-0.1/clean.css" rel="stylesheet" />
<script src="Day1Ex_Final_files/clean-0.1/clean.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    background-color: #ffffff;
    color: #a0a0a0;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #a0a0a0;  padding-left: 4px; }
div.sourceCode
  { color: #1f1c1b; background-color: #ffffff; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span. { color: #1f1c1b; } /* Normal */
code span.al { color: #bf0303; background-color: #f7e6e6; font-weight: bold; } /* Alert */
code span.an { color: #ca60ca; } /* Annotation */
code span.at { color: #0057ae; } /* Attribute */
code span.bn { color: #b08000; } /* BaseN */
code span.bu { color: #644a9b; font-weight: bold; } /* BuiltIn */
code span.cf { color: #1f1c1b; font-weight: bold; } /* ControlFlow */
code span.ch { color: #924c9d; } /* Char */
code span.cn { color: #aa5500; } /* Constant */
code span.co { color: #898887; } /* Comment */
code span.cv { color: #0095ff; } /* CommentVar */
code span.do { color: #607880; } /* Documentation */
code span.dt { color: #0057ae; } /* DataType */
code span.dv { color: #b08000; } /* DecVal */
code span.er { color: #bf0303; text-decoration: underline; } /* Error */
code span.ex { color: #0095ff; font-weight: bold; } /* Extension */
code span.fl { color: #b08000; } /* Float */
code span.fu { color: #644a9b; } /* Function */
code span.im { color: #ff5500; } /* Import */
code span.in { color: #b08000; } /* Information */
code span.kw { color: #1f1c1b; font-weight: bold; } /* Keyword */
code span.op { color: #1f1c1b; } /* Operator */
code span.ot { color: #006e28; } /* Other */
code span.pp { color: #006e28; } /* Preprocessor */
code span.re { color: #0057ae; background-color: #e0e9f8; } /* RegionMarker */
code span.sc { color: #3daee9; } /* SpecialChar */
code span.ss { color: #ff5500; } /* SpecialString */
code span.st { color: #bf0303; } /* String */
code span.va { color: #0057ae; } /* Variable */
code span.vs { color: #bf0303; } /* VerbatimString */
code span.wa { color: #bf0303; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>



<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
</style>
<div class="container-fluid main-container">
  <div class="row">
     <div class="col-md-10">




<div id="header">
<h1 class="title">Day1_1020.1200 - GWAS exercises</h1>
<h4 class="author"><em>Malachy Campbell</em></h4>
<h4 class="date"><em>6/23/2019</em></h4>
</div>

<div id="content">
<p>The goal of this exercise is to introduce participants to popular software for preprocessing SNP data and running GWAS in R. We will use PLINK to preprocess two SNP datasets.</p>
<div id="imputed-rice-data" class="section level2">
<h2>Imputed rice data</h2>
<p>The first dataset consists of imputed SNP data for 1,568 rice accessions. These accessions are from the rice diversity panels (rice diversity panel1, rice diversity panel2), which have been genotyped using 44k SNPs and 700k SNPs. The authors used resequencing data for a larger panel of 3,000 rice accessions to impute more than 5 million SNPs for the 1,568 accessions. Details regarding the accessions and imputation approach are decribed in <a href="https://www.nature.com/articles/s41467-018-05538-1#Abs1">Wang et al (2018)</a>. Given the size of the dataset, we not work directly with this data for the workshop.</p>
<p>I will utilize this data to give participants an idea about how to process large datasets in PLINK. Using PLINK I will (1) extract SNP data for a subset of accessions, (2) prune SNPs that are in high LD, and (3) remove SNPs with low MAF.</p>
<div id="plink" class="section level3">
<h3>PLINK</h3>
<p>PLINK is a set of command line tools that can be used to process SNP data, assess population structure, and perform GWAS. However for the latter two there are many R packages that much more robust than PLINK.</p>
<p>We won’t do any hands on exercises with PLINK. However if you are interested in using PLINK in your own studies, you can download it <a href="https://www.cog-genomics.org/plink/1.9/">here</a>. Once it is downloaded you will need to add the directory to your path. This allows PLINK to be called directly from the command line.</p>
<div id="plink-files" class="section level4">
<h4>PLINK files</h4>
<p>PLINK accepts genotype data in several different formats. The classic format is the PED/fam/map set of files. PED files hold the SNP genotype data and phenotype data. Accessions/individuals are listed in the rows and SNPs are stored in the columns (actually each allele is given its own column). The first six columns are * Family ID * Individual ID * Paternal ID * Maternal ID * Sex (1=male; 2=female; other=unknown) * Phenotype The first two can be used to store the accession IDs and the remaining can be filled with missing data (-9 in PLINK is missing). The FAM files are basically the first six columns of the PED file. The MAP file lists the SNP positions (BP and/or CM), names, and chromosome numbers.</p>
<p>There is also the binary equivalent of the PED/FAM/MAP files, these are BED/FAM/BIM files. They contain exactly the same information but are much more efficient for large datasets.</p>
<p>PLINK also accepts VCF (variant call format) files, 23andMe format, as well as several other formats that are less popular in plant studies.</p>
</div>
<div id="basic-plink-command" class="section level4">
<h4>Basic PLINK command</h4>
<p>Generally what I use PLINK for is processing SNP data. So the workflow is to read some file in, apply some filter, and output the file. In the code below we will read in a binary (BED) formatted file and write a text (PED) file out. We specify that the input data is in binary format using the –bfile argument. If the data was in text format we would use –file.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb1-1" title="1"><span class="ex">plink</span> --bfile RICE_RP_mLIDs --recode --out RICE_RP_text</a></code></pre></div>
<p>–recode will generate a text (PED) formatted file after applying some filters. Here nothing is specified, so it will be the same data as RICE_RP_mLIDs, but in text (PED) format.</p>
</div>
<div id="extracting-snp-data-for-a-subset-of-individuals" class="section level4">
<h4>Extracting SNP data for a subset of individuals</h4>
<p>Here, we will extract SNP data for a subset of accessions that have phenotypes. To do this, you need a text file that lists the accessions you want to subset. This file will be created in R. A quick breakdown of what we are doing:</p>
<ol style="list-style-type: decimal">
<li>Load phenotype file</li>
<li>Find which accessions have phenotype and genotype data</li>
<li>Load the FAM file for all 1,568 accessions and subset for phenotyped and genotyped accessions</li>
<li>Write subsetted FAM file to disk. This will be used as input for PLINK</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1">phenos &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;~/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/VT_Phenotypes.csv&quot;</span>)</a>
<a class="sourceLine" id="cb2-2" title="2">FAM &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;~/Downloads/RICE_RP/RICE_RP_mLIDs.fam&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]</a>
<a class="sourceLine" id="cb2-3" title="3"></a>
<a class="sourceLine" id="cb2-4" title="4">FAM &lt;-<span class="st"> </span>FAM[FAM<span class="op">$</span>V1 <span class="op">%in%</span><span class="st"> </span>phenos<span class="op">$</span>HDRA.ID ,] <span class="co">#365 accessions</span></a>
<a class="sourceLine" id="cb2-5" title="5"></a>
<a class="sourceLine" id="cb2-6" title="6">phenos &lt;-<span class="st"> </span>phenos[phenos<span class="op">$</span>HDRA.ID <span class="op">%in%</span><span class="st"> </span>FAM<span class="op">$</span>V1 ,]</a>
<a class="sourceLine" id="cb2-7" title="7"><span class="kw">write.csv</span>(phenos , <span class="st">&quot;~/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/VT_Phenotypes.csv&quot;</span>, <span class="dt">row.names =</span> F)</a>
<a class="sourceLine" id="cb2-8" title="8"><span class="kw">write.table</span>(FAM, <span class="st">&quot;~/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Accs2Keep.txt&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">col.names =</span> F, <span class="dt">row.names =</span> F, <span class="dt">quote =</span> F)</a></code></pre></div>
<p>This will load the full dataset, extract the SNP data for accesssions in the Accs2Keep.txt file, and create a new set of binary files called VTrice.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb3-1" title="1"><span class="ex">plink</span> --bfile ~/Downloads/RICE_RP/RICE_RP_mLIDs --keep ~/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Accs2Keep.txt --make-bed --out ~/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice</a></code></pre></div>
<p>The output…</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb4-1" title="1"><span class="ex">PLINK</span> v1.90b4.1 64-bit (30 Mar 2017)           <span class="ex">www.cog-genomics.org/plink/1.9/</span></a>
<a class="sourceLine" id="cb4-2" title="2"><span class="kw">(</span><span class="ex">C</span><span class="kw">)</span> <span class="ex">2005-2017</span> Shaun Purcell, Christopher Chang   GNU General Public License v3</a>
<a class="sourceLine" id="cb4-3" title="3"><span class="ex">Logging</span> to /Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice.log.</a>
<a class="sourceLine" id="cb4-4" title="4"><span class="ex">Options</span> in effect:</a>
<a class="sourceLine" id="cb4-5" title="5">  <span class="ex">--bfile</span> /Users/malachycampbell/Downloads/RICE_RP/RICE_RP_mLIDs</a>
<a class="sourceLine" id="cb4-6" title="6">  <span class="ex">--keep</span> /Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Accs2Keep.txt</a>
<a class="sourceLine" id="cb4-7" title="7">  <span class="ex">--make-bed</span></a>
<a class="sourceLine" id="cb4-8" title="8">  <span class="ex">--out</span> /Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice</a>
<a class="sourceLine" id="cb4-9" title="9"></a>
<a class="sourceLine" id="cb4-10" title="10"><span class="ex">16384</span> MB RAM detected<span class="kw">;</span> <span class="ex">reserving</span> 8192 MB for main workspace.</a>
<a class="sourceLine" id="cb4-11" title="11"><span class="ex">5231433</span> variants loaded from .bim file.</a>
<a class="sourceLine" id="cb4-12" title="12"><span class="ex">4591</span> people (0 males, 0 females, 4591 ambiguous) <span class="ex">loaded</span> from .fam.</a>
<a class="sourceLine" id="cb4-13" title="13"><span class="ex">Ambiguous</span> sex IDs written to</a>
<a class="sourceLine" id="cb4-14" title="14"><span class="ex">/Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice.nosex</span></a>
<a class="sourceLine" id="cb4-15" title="15"><span class="ex">.</span></a>
<a class="sourceLine" id="cb4-16" title="16"><span class="ex">--keep</span>: 365 people remaining.</a>
<a class="sourceLine" id="cb4-17" title="17"><span class="ex">Using</span> 1 thread (no multithreaded calculations invoked)<span class="ex">.</span></a>
<a class="sourceLine" id="cb4-18" title="18"><span class="ex">Before</span> main variant filters, 365 founders and 0 nonfounders present.</a>
<a class="sourceLine" id="cb4-19" title="19"><span class="ex">Calculating</span> allele frequencies... done.</a>
<a class="sourceLine" id="cb4-20" title="20"><span class="ex">5231433</span> variants and 365 people pass filters and QC.</a>
<a class="sourceLine" id="cb4-21" title="21"><span class="ex">Note</span>: No phenotypes present.</a>
<a class="sourceLine" id="cb4-22" title="22"><span class="ex">--make-bed</span> to</a>
<a class="sourceLine" id="cb4-23" title="23"><span class="ex">/Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice.bed</span></a>
<a class="sourceLine" id="cb4-24" title="24"><span class="ex">+</span></a>
<a class="sourceLine" id="cb4-25" title="25"><span class="ex">/Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice.bim</span></a>
<a class="sourceLine" id="cb4-26" title="26"><span class="ex">+</span></a>
<a class="sourceLine" id="cb4-27" title="27"><span class="ex">/Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice.fam</span></a>
<a class="sourceLine" id="cb4-28" title="28"><span class="ex">...</span> done.</a></code></pre></div>
</div>
<div id="pruning-the-dataset" class="section level4">
<h4>Pruning the dataset</h4>
<p>Sometimes it is usefull to remove SNPs that have high LD. Here, will find what SNPs are in tight linkage. The indep command takes a window, calculates LD between all SNPs in that window, and removes those that exceed some threshold. The marker with the higher MAF is kept. PLINK uses the variance inflation factor (VIF) <span class="math inline">\(VIF = \frac{1}{1 - r^2}\)</span> as a metric for linkage. I will remove SNPs that are in very tight linkage (<span class="math inline">\(r \geq 0.99\)</span>). So the VIF will be <span class="math inline">\(VIF = \frac{1}{1 - 0.80^2} = 2.777778\)</span>.</p>
<p>It is important to note that PLINK does not do any filtering when the indep arguement is used. PLINK will output two files that list the SNPs with <span class="math inline">\(VIF &lt; thresh\)</span> in the .in file and <span class="math inline">\(VIF \geq thresh\)</span> in the .out file.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb5-1" title="1"><span class="ex">plink</span> --bfile ~/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice --indep 10 5 2.777778 --make-bed --out ~/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_pruned</a></code></pre></div>
<p>The output…</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb6-1" title="1"><span class="ex">PLINK</span> v1.90b4.1 64-bit (30 Mar 2017)           <span class="ex">www.cog-genomics.org/plink/1.9/</span></a>
<a class="sourceLine" id="cb6-2" title="2"><span class="kw">(</span><span class="ex">C</span><span class="kw">)</span> <span class="ex">2005-2017</span> Shaun Purcell, Christopher Chang   GNU General Public License v3</a>
<a class="sourceLine" id="cb6-3" title="3"><span class="ex">Logging</span> to /Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_pruned.log.</a>
<a class="sourceLine" id="cb6-4" title="4"><span class="ex">Options</span> in effect:</a>
<a class="sourceLine" id="cb6-5" title="5">  <span class="ex">--bfile</span> /Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice</a>
<a class="sourceLine" id="cb6-6" title="6">  <span class="ex">--indep</span> 10 5 2.777778</a>
<a class="sourceLine" id="cb6-7" title="7">  <span class="ex">--make-bed</span></a>
<a class="sourceLine" id="cb6-8" title="8">  <span class="ex">--out</span> /Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_pruned</a>
<a class="sourceLine" id="cb6-9" title="9"></a>
<a class="sourceLine" id="cb6-10" title="10"><span class="ex">16384</span> MB RAM detected<span class="kw">;</span> <span class="ex">reserving</span> 8192 MB for main workspace.</a>
<a class="sourceLine" id="cb6-11" title="11"><span class="ex">5231433</span> variants loaded from .bim file.</a>
<a class="sourceLine" id="cb6-12" title="12"><span class="ex">365</span> people (0 males, 0 females, 365 ambiguous) <span class="ex">loaded</span> from .fam.</a>
<a class="sourceLine" id="cb6-13" title="13"><span class="ex">Ambiguous</span> sex IDs written to</a>
<a class="sourceLine" id="cb6-14" title="14"><span class="ex">/Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_pruned.nosex</span></a>
<a class="sourceLine" id="cb6-15" title="15"><span class="ex">.</span></a>
<a class="sourceLine" id="cb6-16" title="16"><span class="ex">Using</span> 1 thread (no multithreaded calculations invoked)<span class="ex">.</span></a>
<a class="sourceLine" id="cb6-17" title="17"><span class="ex">Before</span> main variant filters, 365 founders and 0 nonfounders present.</a>
<a class="sourceLine" id="cb6-18" title="18"><span class="ex">Calculating</span> allele frequencies... done.</a>
<a class="sourceLine" id="cb6-19" title="19"><span class="ex">5231433</span> variants and 365 people pass filters and QC.</a>
<a class="sourceLine" id="cb6-20" title="20"><span class="ex">Note</span>: No phenotypes present.</a>
<a class="sourceLine" id="cb6-21" title="21"><span class="ex">--make-bed</span> to</a>
<a class="sourceLine" id="cb6-22" title="22"><span class="ex">/Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_pruned.bed</span></a>
<a class="sourceLine" id="cb6-23" title="23"><span class="ex">+</span></a>
<a class="sourceLine" id="cb6-24" title="24"><span class="ex">/Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_pruned.bim</span></a>
<a class="sourceLine" id="cb6-25" title="25"><span class="ex">+</span></a>
<a class="sourceLine" id="cb6-26" title="26"><span class="ex">/Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_pruned.fam</span></a>
<a class="sourceLine" id="cb6-27" title="27"><span class="ex">...</span> done.</a>
<a class="sourceLine" id="cb6-28" title="28"><span class="ex">Pruned</span> 373200 variants from chromosome 1, leaving 248680.</a>
<a class="sourceLine" id="cb6-29" title="29"><span class="ex">Pruned</span> 279799 variants from chromosome 2, leaving 193116.</a>
<a class="sourceLine" id="cb6-30" title="30"><span class="ex">Pruned</span> 265348 variants from chromosome 3, leaving 174968.</a>
<a class="sourceLine" id="cb6-31" title="31"><span class="ex">Pruned</span> 265869 variants from chromosome 4, leaving 196661.</a>
<a class="sourceLine" id="cb6-32" title="32"><span class="ex">Pruned</span> 230390 variants from chromosome 5, leaving 148659.</a>
<a class="sourceLine" id="cb6-33" title="33"><span class="ex">Pruned</span> 287508 variants from chromosome 6, leaving 191051.</a>
<a class="sourceLine" id="cb6-34" title="34"><span class="ex">Pruned</span> 249895 variants from chromosome 7, leaving 165879.</a>
<a class="sourceLine" id="cb6-35" title="35"><span class="ex">Pruned</span> 243153 variants from chromosome 8, leaving 198439.</a>
<a class="sourceLine" id="cb6-36" title="36"><span class="ex">Pruned</span> 179336 variants from chromosome 9, leaving 132674.</a>
<a class="sourceLine" id="cb6-37" title="37"><span class="ex">Pruned</span> 204244 variants from chromosome 10, leaving 144218.</a>
<a class="sourceLine" id="cb6-38" title="38"><span class="ex">Pruned</span> 254068 variants from chromosome 11, leaving 215808.</a>
<a class="sourceLine" id="cb6-39" title="39"><span class="ex">Pruned</span> 222199 variants from chromosome 12, leaving 166271.</a>
<a class="sourceLine" id="cb6-40" title="40"><span class="ex">Pruning</span> complete.  3055009 of 5231433 variants removed.</a>
<a class="sourceLine" id="cb6-41" title="41"><span class="ex">Marker</span> lists written to</a>
<a class="sourceLine" id="cb6-42" title="42"><span class="ex">/Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_pruned.prune.in</span></a>
<a class="sourceLine" id="cb6-43" title="43"><span class="ex">and</span></a>
<a class="sourceLine" id="cb6-44" title="44"><span class="ex">/Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_pruned.prune.out</span></a>
<a class="sourceLine" id="cb6-45" title="45"><span class="ex">.</span></a></code></pre></div>
<p>Now we will extract the SNPs that do not exceed that threshold and remove SNPs with a MAF <span class="math inline">\(\leq 0.05\)</span></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb7-1" title="1"><span class="ex">plink</span> --bfile /Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice --maf 0.05 --extract /Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_pruned.prune.in --make-bed --out /Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_thinned_maf</a></code></pre></div>
<p>The output…</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb8-1" title="1"><span class="ex">PLINK</span> v1.90b4.1 64-bit (30 Mar 2017)           <span class="ex">www.cog-genomics.org/plink/1.9/</span></a>
<a class="sourceLine" id="cb8-2" title="2"><span class="kw">(</span><span class="ex">C</span><span class="kw">)</span> <span class="ex">2005-2017</span> Shaun Purcell, Christopher Chang   GNU General Public License v3</a>
<a class="sourceLine" id="cb8-3" title="3"><span class="ex">Logging</span> to /Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_thinned_maf.log.</a>
<a class="sourceLine" id="cb8-4" title="4"><span class="ex">Options</span> in effect:</a>
<a class="sourceLine" id="cb8-5" title="5">  <span class="ex">--bfile</span> /Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice</a>
<a class="sourceLine" id="cb8-6" title="6">  <span class="ex">--extract</span> /Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_pruned.prune.in</a>
<a class="sourceLine" id="cb8-7" title="7">  <span class="ex">--maf</span> 0.05</a>
<a class="sourceLine" id="cb8-8" title="8">  <span class="ex">--make-bed</span></a>
<a class="sourceLine" id="cb8-9" title="9">  <span class="ex">--out</span> /Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_thinned_maf</a>
<a class="sourceLine" id="cb8-10" title="10">  <span class="ex">--recode</span></a>
<a class="sourceLine" id="cb8-11" title="11"></a>
<a class="sourceLine" id="cb8-12" title="12"><span class="ex">16384</span> MB RAM detected<span class="kw">;</span> <span class="ex">reserving</span> 8192 MB for main workspace.</a>
<a class="sourceLine" id="cb8-13" title="13"><span class="ex">5231433</span> variants loaded from .bim file.</a>
<a class="sourceLine" id="cb8-14" title="14"><span class="ex">365</span> people (0 males, 0 females, 365 ambiguous) <span class="ex">loaded</span> from .fam.</a>
<a class="sourceLine" id="cb8-15" title="15"><span class="ex">Ambiguous</span> sex IDs written to</a>
<a class="sourceLine" id="cb8-16" title="16"><span class="ex">/Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_thinned_maf.nosex</span></a>
<a class="sourceLine" id="cb8-17" title="17"><span class="ex">.</span></a>
<a class="sourceLine" id="cb8-18" title="18"><span class="ex">--extract</span>: 2176424 variants remaining.</a>
<a class="sourceLine" id="cb8-19" title="19"><span class="ex">Using</span> 1 thread (no multithreaded calculations invoked)<span class="ex">.</span></a>
<a class="sourceLine" id="cb8-20" title="20"><span class="ex">Before</span> main variant filters, 365 founders and 0 nonfounders present.</a>
<a class="sourceLine" id="cb8-21" title="21"><span class="ex">Calculating</span> allele frequencies... done.</a>
<a class="sourceLine" id="cb8-22" title="22"><span class="ex">983225</span> variants removed due to minor allele threshold(s)</a>
<a class="sourceLine" id="cb8-23" title="23"><span class="kw">(</span><span class="ex">--maf/--max-maf/--mac/--max-mac</span><span class="kw">)</span><span class="ex">.</span></a>
<a class="sourceLine" id="cb8-24" title="24"><span class="ex">1193199</span> variants and 365 people pass filters and QC.</a>
<a class="sourceLine" id="cb8-25" title="25"><span class="ex">Note</span>: No phenotypes present.</a>
<a class="sourceLine" id="cb8-26" title="26"><span class="ex">--make-bed</span> to</a>
<a class="sourceLine" id="cb8-27" title="27"><span class="ex">/Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_thinned_maf.bed</span></a>
<a class="sourceLine" id="cb8-28" title="28"><span class="ex">+</span></a>
<a class="sourceLine" id="cb8-29" title="29"><span class="ex">/Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_thinned_maf.bim</span></a>
<a class="sourceLine" id="cb8-30" title="30"><span class="ex">+</span></a>
<a class="sourceLine" id="cb8-31" title="31"><span class="ex">/Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_thinned_maf.fam</span></a>
<a class="sourceLine" id="cb8-32" title="32"><span class="ex">...</span> done.</a>
<a class="sourceLine" id="cb8-33" title="33"><span class="ex">--recode</span> ped to</a>
<a class="sourceLine" id="cb8-34" title="34"><span class="ex">/Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_thinned_maf.ped</span></a>
<a class="sourceLine" id="cb8-35" title="35"><span class="ex">+</span></a>
<a class="sourceLine" id="cb8-36" title="36"><span class="ex">/Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_thinned_maf.map</span></a>
<a class="sourceLine" id="cb8-37" title="37"><span class="ex">...</span> done.</a></code></pre></div>
<p>This is the dataset that is made available to everyone.</p>
</div>
</div>
</div>
<div id="gwas" class="section level2">
<h2>GWAS</h2>
<p>I will demonstrate the use of two different R packages for running GWAS: GAPIT and rrBLUP. These two packages essentially run the same models, but offer more or less options. Most of the work is in formatting the input files. The downside of these packages is they assume there is only a single observation for each individual. So the workflow for these (which we will not discuss in this workshop) is to calculate best linear unbiased estimates (BLUEs) or LSMeans from the raw phenotype data and to use this as an input to these packages. So basically a two-step approach.</p>
<p>rrBLUP and sommer are maintained by CRAN, GAPIT is not.</p>
<div id="rrblup" class="section level3">
<h3>rrBLUP</h3>
<p>The first step is formatting the data correctly. We can run a mixed model for GWAS with the rrBLUP function GWAS.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"><span class="co">#install.packages(&quot;rrBLUP&quot;) #install it if you haven&#39;t already</span></a>
<a class="sourceLine" id="cb9-2" title="2"><span class="kw">library</span>(rrBLUP) <span class="co">#load the package</span></a>
<a class="sourceLine" id="cb9-3" title="3"></a>
<a class="sourceLine" id="cb9-4" title="4">?GWAS <span class="co">#get help for the function</span></a></code></pre></div>
<div id="formatting-snp-data-for-rrblup" class="section level4">
<h4>Formatting SNP data for rrBLUP</h4>
<p>rrBLUP accepts genotypic data as dataframe. The first three columns are: marker names, chromosome and position. The subsequent columns list the SNP genotypes in numeric format. Here, aa = -1, Aa = 0, and AA = 1. It doesn’t matter which allele (major or minor) is considered A or a.</p>
<p>To format the SNP data will will use the GASTON package. The bed file can be read in using the read.bed.matrix function. All you need to do is supply the path and basename of the PLINK files</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"><span class="co">#install.packages(&quot;gaston&quot;) #install it if you haven&#39;t already</span></a>
<a class="sourceLine" id="cb10-2" title="2"><span class="kw">library</span>(gaston)</a></code></pre></div>
<pre><code>## Loading required package: Rcpp</code></pre>
<pre><code>## Loading required package: RcppParallel</code></pre>
<pre><code>## 
## Attaching package: &#39;RcppParallel&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:Rcpp&#39;:
## 
##     LdFlags</code></pre>
<pre><code>## Gaston set number of threads to 4. Use setThreadOptions() to modify this.</code></pre>
<pre><code>## 
## Attaching package: &#39;gaston&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:stats&#39;:
## 
##     sigma</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     cbind, rbind</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1">SNPs &lt;-<span class="st"> </span><span class="kw">read.bed.matrix</span>(<span class="dt">basename =</span> <span class="st">&quot;/Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_thinned_maf&quot;</span>)</a></code></pre></div>
<pre><code>## Reading /Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_thinned_maf.fam 
## Reading /Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_thinned_maf.bim 
## Reading /Users/malachycampbell/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Genotypes/VTrice_thinned_maf.bed 
## ped stats and snps stats have been set. 
## &#39;p&#39; has been set. 
## &#39;mu&#39; and &#39;sigma&#39; have been set.</code></pre>
<p>GASTON doesn’t hold the marker data in memory, so it is nice for big datasets. However, it holds all the other information in the FAM and MAP file. We can check out what is stored in SNPs.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" title="1"><span class="kw">str</span>(SNPs)</a></code></pre></div>
<pre><code>## Formal class &#39;bed.matrix&#39; [package &quot;gaston&quot;] with 8 slots
##   ..@ ped                 :&#39;data.frame&#39;: 365 obs. of  30 variables:
##   .. ..$ famid      : chr [1:365] &quot;006dfe9b.0&quot; &quot;02cc7c6d.0&quot; &quot;02d095ba.0&quot; &quot;07dac217.0&quot; ...
##   .. ..$ id         : chr [1:365] &quot;006dfe9b.0&quot; &quot;02cc7c6d.0&quot; &quot;02d095ba.0&quot; &quot;07dac217.0&quot; ...
##   .. ..$ father     : int [1:365] 0 0 0 0 0 0 0 0 0 0 ...
##   .. ..$ mother     : int [1:365] 0 0 0 0 0 0 0 0 0 0 ...
##   .. ..$ sex        : int [1:365] 0 0 0 0 0 0 0 0 0 0 ...
##   .. ..$ pheno      : int [1:365] -9 -9 -9 -9 -9 -9 -9 -9 -9 -9 ...
##   .. ..$ N0         : int [1:365] 337396 316801 213361 130013 333728 133546 334592 122923 120716 280561 ...
##   .. ..$ N1         : int [1:365] 22981 12799 12342 6326 13490 8496 19812 6573 6539 22718 ...
##   .. ..$ N2         : int [1:365] 832822 863599 967496 1056860 845981 1051157 838795 1063703 1065944 889920 ...
##   .. ..$ NAs        : int [1:365] 0 0 0 0 0 0 0 0 0 0 ...
##   .. ..$ N0.x       : int [1:365] 0 0 0 0 0 0 0 0 0 0 ...
##   .. ..$ N1.x       : int [1:365] 0 0 0 0 0 0 0 0 0 0 ...
##   .. ..$ N2.x       : int [1:365] 0 0 0 0 0 0 0 0 0 0 ...
##   .. ..$ NAs.x      : int [1:365] 0 0 0 0 0 0 0 0 0 0 ...
##   .. ..$ N0.y       : int [1:365] 0 0 0 0 0 0 0 0 0 0 ...
##   .. ..$ N1.y       : int [1:365] 0 0 0 0 0 0 0 0 0 0 ...
##   .. ..$ N2.y       : int [1:365] 0 0 0 0 0 0 0 0 0 0 ...
##   .. ..$ NAs.y      : int [1:365] 0 0 0 0 0 0 0 0 0 0 ...
##   .. ..$ N0.mt      : int [1:365] 0 0 0 0 0 0 0 0 0 0 ...
##   .. ..$ N1.mt      : int [1:365] 0 0 0 0 0 0 0 0 0 0 ...
##   .. ..$ N2.mt      : int [1:365] 0 0 0 0 0 0 0 0 0 0 ...
##   .. ..$ NAs.mt     : int [1:365] 0 0 0 0 0 0 0 0 0 0 ...
##   .. ..$ callrate   : num [1:365] 1 1 1 1 1 1 1 1 1 1 ...
##   .. ..$ hz         : num [1:365] 0.0193 0.0107 0.0103 0.0053 0.0113 ...
##   .. ..$ callrate.x : num [1:365] NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN ...
##   .. ..$ hz.x       : num [1:365] NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN ...
##   .. ..$ callrate.y : num [1:365] NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN ...
##   .. ..$ hz.y       : num [1:365] NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN ...
##   .. ..$ callrate.mt: num [1:365] NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN ...
##   .. ..$ hz.mt      : num [1:365] NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN ...
##   ..@ snps                :&#39;data.frame&#39;: 1193199 obs. of  17 variables:
##   .. ..$ chr     : int [1:1193199] 1 1 1 1 1 1 1 1 1 1 ...
##   .. ..$ id      : chr [1:1193199] &quot;mlid0000000099&quot; &quot;mlid0000000110&quot; &quot;mlid0000000148&quot; &quot;mlid0000000166&quot; ...
##   .. ..$ dist    : int [1:1193199] 0 0 0 0 0 0 0 0 0 0 ...
##   .. ..$ pos     : int [1:1193199] 1151 1178 1266 1297 1325 1573 1579 1695 1733 2033 ...
##   .. ..$ A1      : chr [1:1193199] &quot;A&quot; &quot;T&quot; &quot;A&quot; &quot;A&quot; ...
##   .. ..$ A2      : chr [1:1193199] &quot;C&quot; &quot;G&quot; &quot;G&quot; &quot;G&quot; ...
##   .. ..$ N0      : int [1:1193199] 19 56 44 56 44 56 23 39 58 57 ...
##   .. ..$ N1      : int [1:1193199] 6 3 6 3 7 3 2 9 2 2 ...
##   .. ..$ N2      : int [1:1193199] 340 306 315 306 314 306 340 317 305 306 ...
##   .. ..$ NAs     : int [1:1193199] 0 0 0 0 0 0 0 0 0 0 ...
##   .. ..$ N0.f    : int [1:1193199] NA NA NA NA NA NA NA NA NA NA ...
##   .. ..$ N1.f    : int [1:1193199] NA NA NA NA NA NA NA NA NA NA ...
##   .. ..$ N2.f    : int [1:1193199] NA NA NA NA NA NA NA NA NA NA ...
##   .. ..$ NAs.f   : int [1:1193199] NA NA NA NA NA NA NA NA NA NA ...
##   .. ..$ callrate: num [1:1193199] 1 1 1 1 1 1 1 1 1 1 ...
##   .. ..$ maf     : num [1:1193199] 0.0603 0.1575 0.1288 0.1575 0.1301 ...
##   .. ..$ hz      : num [1:1193199] 0.01644 0.00822 0.01644 0.00822 0.01918 ...
##   ..@ bed                 :&lt;externalptr&gt; 
##   ..@ p                   : num [1:1193199] 0.94 0.842 0.871 0.842 0.87 ...
##   ..@ mu                  : num [1:1193199] 1.88 1.68 1.74 1.68 1.74 ...
##   ..@ sigma               : num [1:1193199] 0.459 0.724 0.658 0.724 0.659 ...
##   ..@ standardize_p       : logi FALSE
##   ..@ standardize_mu_sigma: logi FALSE</code></pre>
<p>We can perform filtering on the SNPs, even without holding the markers in memory. However, these steps aren’t necessary since we already did our filtering in PLINK.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" title="1">SNPs &lt;-<span class="st"> </span><span class="kw">select.snps</span>(SNPs, maf <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.05</span>)</a>
<a class="sourceLine" id="cb23-2" title="2"><span class="kw">dim</span>(SNPs)</a>
<a class="sourceLine" id="cb23-3" title="3">SNPs &lt;-<span class="st"> </span><span class="kw">select.snps</span>(SNPs, callrate <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.9</span>)</a>
<a class="sourceLine" id="cb23-4" title="4"><span class="kw">dim</span>(SNPs)</a></code></pre></div>
<p>Let’s extract the marker data and format it for rrBLUP.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" title="1">genos &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(SNPs)</a>
<a class="sourceLine" id="cb24-2" title="2"><span class="kw">row.names</span>(genos) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;X&quot;</span>, <span class="kw">row.names</span>(genos))</a>
<a class="sourceLine" id="cb24-3" title="3"><span class="kw">dim</span>(genos) <span class="co"># A BIG n x m matrix</span></a></code></pre></div>
<pre><code>## [1]     365 1193199</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" title="1"><span class="kw">head</span>(genos[,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]) <span class="co">#SNPs are coded as 0, 1, 2</span></a></code></pre></div>
<pre><code>##             mlid0000000099 mlid0000000110 mlid0000000148
## X006dfe9b.0              2              0              2
## X02cc7c6d.0              2              2              2
## X02d095ba.0              2              2              2
## X07dac217.0              0              2              2
## X07f246bb.0              2              2              0
## X08de34ee.0              2              2              2</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" title="1">genos &lt;-<span class="st"> </span>genos <span class="op">-</span><span class="st"> </span><span class="dv">1</span> <span class="co">#Now they are coded as -1, 0, 1</span></a>
<a class="sourceLine" id="cb28-2" title="2"></a>
<a class="sourceLine" id="cb28-3" title="3"><span class="co">#Recall rrBLUP likes the genotypes to be in a dataframe with the first 3 columns giving the marker information</span></a>
<a class="sourceLine" id="cb28-4" title="4">genos &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">rs =</span> SNPs<span class="op">@</span>snps<span class="op">$</span>id, <span class="dt">chr =</span> SNPs<span class="op">@</span>snps<span class="op">$</span>chr, <span class="dt">bp =</span> SNPs<span class="op">@</span>snps<span class="op">$</span>pos, <span class="kw">t</span>(genos))</a>
<a class="sourceLine" id="cb28-5" title="5"></a>
<a class="sourceLine" id="cb28-6" title="6"><span class="co">#Since this is a huge dataset it will take a very long time to run. For the demonstration purposes we will randomly select a subset of 50,000 SNPs.</span></a>
<a class="sourceLine" id="cb28-7" title="7"><span class="kw">set.seed</span>(<span class="dv">123</span>)</a>
<a class="sourceLine" id="cb28-8" title="8">genoIndx &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="kw">nrow</span>(genos), <span class="dt">size =</span> <span class="dv">50000</span>, <span class="dt">replace =</span> F)</a>
<a class="sourceLine" id="cb28-9" title="9"></a>
<a class="sourceLine" id="cb28-10" title="10">genos &lt;-<span class="st"> </span>genos[genoIndx ,]</a>
<a class="sourceLine" id="cb28-11" title="11"><span class="kw">table</span>(genos<span class="op">$</span>chr)</a></code></pre></div>
<pre><code>## 
##    1    2    3    4    5    6    7    8    9   10   11   12 
## 6004 4320 4127 4212 3301 4517 3676 4429 2960 3866 4676 3912</code></pre>
</div>
<div id="formatting-phenotypes-for-rrblup" class="section level4">
<h4>Formatting phenotypes for rrBLUP</h4>
<p>The phenotypes should be stored in a dataframe where the first column in the “gid”, which should match the columns of genos (4th column on).</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" title="1">phenos &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;~/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/VT_Phenotypes.csv&quot;</span>)</a>
<a class="sourceLine" id="cb30-2" title="2">phenos &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">gid =</span> <span class="kw">paste0</span>(<span class="st">&quot;X&quot;</span>, phenos<span class="op">$</span>HDRA.ID), <span class="dt">Y =</span> phenos<span class="op">$</span>Na.Root.LSMEAN)</a>
<a class="sourceLine" id="cb30-3" title="3"><span class="kw">head</span>(phenos)</a></code></pre></div>
<pre><code>##           gid         Y
## 1 X86f75d2b.0 0.2933769
## 2 X2e1c9c87.0 0.2970459
## 3 X1f10be3d.0 0.3127060
## 4 X9f782e9e.0 0.3068186
## 5 X6e26f4cc.0 0.4205350
## 6 X8c76404c.0 0.2723217</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" title="1">phenos &lt;-<span class="st"> </span>phenos[<span class="kw">match</span>(<span class="kw">colnames</span>(genos)[<span class="dv">4</span><span class="op">:</span><span class="kw">ncol</span>(genos)], phenos<span class="op">$</span>gid) ,]</a></code></pre></div>
</div>
<div id="running-gwas-with-rrblup" class="section level4">
<h4>Running GWAS with rrBLUP</h4>
<p>Running GWAS is pretty simple. rrBLUP will calculate the kinship matrix for you using their A.mat function. Moreover, if you already know the number of PC’s you want to include in your model, then there is no need to supply the function with PCs. rrBLUP will do this for you.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" title="1">Rt.Na.GWAS &lt;-<span class="st"> </span><span class="kw">GWAS</span>(<span class="dt">pheno =</span> phenos, <span class="dt">geno =</span> genos, <span class="dt">fixed =</span> <span class="ot">NULL</span>, <span class="dt">K =</span> <span class="ot">NULL</span>, <span class="dt">n.PC =</span> <span class="dv">0</span>,</a>
<a class="sourceLine" id="cb33-2" title="2">     <span class="dt">min.MAF=</span><span class="fl">0.05</span>, <span class="dt">n.core =</span> <span class="dv">1</span>, <span class="dt">P3D =</span> <span class="ot">TRUE</span>, <span class="dt">plot =</span> F)</a></code></pre></div>
<pre><code>## [1] &quot;GWAS for trait: Y&quot;
## [1] &quot;Variance components estimated. Testing markers.&quot;</code></pre>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" title="1"><span class="co">#The kinship matrix will be estimated by rrBLUP. Alternatively we can supply our own.</span></a>
<a class="sourceLine" id="cb35-2" title="2"><span class="co">#P3D = T means that the variance components will be estimated once using a model without any SNP effects. Then these values will be used as input for all other SNPs.</span></a>
<a class="sourceLine" id="cb35-3" title="3"><span class="co">#n.PC: the number of PCs. From the lecture it seems that population structure was adequately accounted for using just the kinship matrix</span></a>
<a class="sourceLine" id="cb35-4" title="4"></a>
<a class="sourceLine" id="cb35-5" title="5"><span class="kw">head</span>(Rt.Na.GWAS) <span class="co"># The output is a dataframe with the marker name, chromosome, postion and -log10(p) value</span></a></code></pre></div>
<pre><code>##                            rs chr    bp         Y
## mlid0000002064 mlid0000002064   1 14823 0.6997230
## mlid0000004235 mlid0000004235   1 30187 0.2888852
## mlid0000004358 mlid0000004358   1 31478 1.8697770
## mlid0000004878 mlid0000004878   1 35240 0.3946891
## mlid0000005926 mlid0000005926   1 43786 0.8621876
## mlid0000007049 mlid0000007049   1 48236 0.2218073</code></pre>
</div>
<div id="plotting-the-results" class="section level4">
<h4>Plotting the results</h4>
<p>The GASTON package provides some nice functions for plotting the results from GWAS analysis.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb37-1" title="1">Rt.Na.GWAS<span class="op">$</span>p &lt;-<span class="st"> </span><span class="dv">10</span><span class="op">^</span>(<span class="op">-</span>Rt.Na.GWAS<span class="op">$</span>Y)</a>
<a class="sourceLine" id="cb37-2" title="2"><span class="kw">colnames</span>(Rt.Na.GWAS)[<span class="dv">3</span>] &lt;-<span class="st"> &quot;pos&quot;</span></a>
<a class="sourceLine" id="cb37-3" title="3"><span class="kw">manhattan</span>(Rt.Na.GWAS)</a></code></pre></div>
<p><img src="Day1Ex_Final_files/figure-html/unnamed-chunk-17-1.png" width="576" /></p>
<p><img src="Day1Ex_Final_files/figure-html/unnamed-chunk-18-1.png" width=".49\linewidth" style="display: block; margin: auto;" /></p>
<p><img src="Day1Ex_Final_files/figure-html/unnamed-chunk-19-1.png" width=".49\linewidth" style="display: block; margin: auto;" /></p>
<p>Lets write the phenotype and genotype data to disk for future use.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" title="1"><span class="kw">write.table</span>(genos, <span class="st">&quot;~/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/NumericGenos.txt&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">col.names =</span> T, <span class="dt">row.names =</span> F, <span class="dt">quote =</span> F)</a>
<a class="sourceLine" id="cb38-2" title="2"></a>
<a class="sourceLine" id="cb38-3" title="3"><span class="kw">write.csv</span>(phenos, <span class="st">&quot;~/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Phenotypes.csv&quot;</span>, <span class="dt">row.names =</span> F)</a></code></pre></div>
</div>
</div>
<div id="gapit" class="section level3">
<h3>GAPIT</h3>
<p>GAPIT is a set of tools for GWAS and genomic prediction. It is quite useful and provides a number of diagonostic plots for GWAS. Like rrBLUP a lot of the work is done behind the scenes. The GAPIT documentation is available <a href="http://www.zzlab.net/GAPIT/">here</a>.</p>
<div id="installing-the-prerequisites" class="section level4">
<h4>Installing the prerequisites</h4>
<p>Since this is not maintained by CRAN, we will have to install all the prerequisites and download the functions from their website.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" title="1"><span class="co">#if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))</span></a>
<a class="sourceLine" id="cb39-2" title="2"><span class="co">#     install.packages(&quot;BiocManager&quot;)</span></a>
<a class="sourceLine" id="cb39-3" title="3"><span class="co">#BiocManager::install(&quot;multtest&quot;)</span></a>
<a class="sourceLine" id="cb39-4" title="4"></a>
<a class="sourceLine" id="cb39-5" title="5"><span class="co">#install.packages(&quot;gplots&quot;)</span></a>
<a class="sourceLine" id="cb39-6" title="6"><span class="co">#install.packages(&quot;LDheatmap&quot;)</span></a>
<a class="sourceLine" id="cb39-7" title="7"><span class="co">#install.packages(&quot;genetics&quot;)</span></a>
<a class="sourceLine" id="cb39-8" title="8"><span class="co">#install.packages(&quot;MASS&quot;)</span></a>
<a class="sourceLine" id="cb39-9" title="9"><span class="co">##install.packages(&quot;compiler&quot;)</span></a>
<a class="sourceLine" id="cb39-10" title="10"><span class="co">#install.packages(&quot;scatterplot3d&quot;)</span></a>
<a class="sourceLine" id="cb39-11" title="11"></a>
<a class="sourceLine" id="cb39-12" title="12"><span class="kw">library</span>(multtest)</a></code></pre></div>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:parallel&#39;:
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from &#39;package:gaston&#39;:
## 
##     cbind, rbind</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, append, as.data.frame, basename, cbind,
##     colMeans, colnames, colSums, dirname, do.call, duplicated,
##     eval, evalq, Filter, Find, get, grep, grepl, intersect,
##     is.unsorted, lapply, lengths, Map, mapply, match, mget, order,
##     paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind,
##     Reduce, rowMeans, rownames, rowSums, sapply, setdiff, sort,
##     table, tapply, union, unique, unsplit, which, which.max,
##     which.min</code></pre>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     &#39;browseVignettes()&#39;. To cite Bioconductor, see
##     &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb49-1" title="1"><span class="kw">library</span>(gplots)</a></code></pre></div>
<pre><code>## 
## Attaching package: &#39;gplots&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:multtest&#39;:
## 
##     wapply</code></pre>
<pre><code>## The following object is masked from &#39;package:stats&#39;:
## 
##     lowess</code></pre>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb53-1" title="1"><span class="kw">library</span>(LDheatmap)</a>
<a class="sourceLine" id="cb53-2" title="2"><span class="kw">library</span>(genetics)</a></code></pre></div>
<pre><code>## Loading required package: combinat</code></pre>
<pre><code>## 
## Attaching package: &#39;combinat&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:utils&#39;:
## 
##     combn</code></pre>
<pre><code>## Loading required package: gdata</code></pre>
<pre><code>## gdata: read.xls support for &#39;XLS&#39; (Excel 97-2004) files ENABLED.</code></pre>
<pre><code>## </code></pre>
<pre><code>## gdata: read.xls support for &#39;XLSX&#39; (Excel 2007+) files ENABLED.</code></pre>
<pre><code>## 
## Attaching package: &#39;gdata&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:Biobase&#39;:
## 
##     combine</code></pre>
<pre><code>## The following object is masked from &#39;package:BiocGenerics&#39;:
## 
##     combine</code></pre>
<pre><code>## The following object is masked from &#39;package:stats&#39;:
## 
##     nobs</code></pre>
<pre><code>## The following object is masked from &#39;package:utils&#39;:
## 
##     object.size</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     startsWith</code></pre>
<pre><code>## Loading required package: gtools</code></pre>
<pre><code>## Loading required package: MASS</code></pre>
<pre><code>## Loading required package: mvtnorm</code></pre>
<pre><code>## </code></pre>
<pre><code>## NOTE: THIS PACKAGE IS NOW OBSOLETE.</code></pre>
<pre><code>## </code></pre>
<pre><code>##   The R-Genetics project has developed an set of enhanced genetics</code></pre>
<pre><code>##   packages to replace &#39;genetics&#39;. Please visit the project homepage</code></pre>
<pre><code>##   at http://rgenetics.org for informtion.</code></pre>
<pre><code>## </code></pre>
<pre><code>## 
## Attaching package: &#39;genetics&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:BiocGenerics&#39;:
## 
##     order</code></pre>
<pre><code>## The following object is masked from &#39;package:gaston&#39;:
## 
##     LD</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     %in%, as.factor, order</code></pre>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb81-1" title="1"><span class="kw">library</span>(MASS)</a>
<a class="sourceLine" id="cb81-2" title="2"><span class="kw">library</span>(compiler)</a>
<a class="sourceLine" id="cb81-3" title="3"><span class="kw">library</span>(scatterplot3d)</a>
<a class="sourceLine" id="cb81-4" title="4"></a>
<a class="sourceLine" id="cb81-5" title="5"><span class="kw">setwd</span>(<span class="st">&quot;~/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/GAPIT&quot;</span>)</a>
<a class="sourceLine" id="cb81-6" title="6"></a>
<a class="sourceLine" id="cb81-7" title="7"><span class="kw">source</span>(<span class="st">&quot;http://zzlab.net/GAPIT/emma.txt&quot;</span>)</a>
<a class="sourceLine" id="cb81-8" title="8"><span class="kw">source</span>(<span class="st">&quot;http://zzlab.net/GAPIT/gapit_functions.txt&quot;</span>)</a></code></pre></div>
<pre><code>## Loading required package: ape</code></pre>
<pre><code>## Loading required package: EMMREML</code></pre>
<pre><code>## Loading required package: Matrix</code></pre>
</div>
<div id="input-for-gapit" class="section level4">
<h4>Input for GAPIT</h4>
<p>GAPIT accepts SNP data in two formats: HapMap and numeric (0,1,2 for aa, Aa, and AA). In almost all cases it will be much more convienient to convert your SNP data to numeric format as above.</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb85-1" title="1">genos &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;~/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/NumericGenos.txt&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">header =</span> T)</a>
<a class="sourceLine" id="cb85-2" title="2">genos &lt;-<span class="st"> </span>genos[<span class="kw">order</span>(genos<span class="op">$</span>chr, genos<span class="op">$</span>bp) ,]</a>
<a class="sourceLine" id="cb85-3" title="3">phenos &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;~/Documents/Dropbox/Work/Presentations/VT_GWAS_workshop/Day1Ex/Phenotypes.csv&quot;</span>)</a></code></pre></div>
<p>When using numeric genotype data, GAPIT requires two objects: (1) the MAP file (chromosome, position) and (2) the genotypes (0,1,2 format). In the GAPIT function we specify these objects with the GM and GD arguments, respectively. GD has markers in the columns and accessions in the rows. This information is already contained in the genos object.</p>
<p>The phenotypes are stored in a dataframe where the first column is the accession id’s and the subsequent columns are the phenotypes.</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb86-1" title="1">myGM &lt;-<span class="st"> </span>genos[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a>
<a class="sourceLine" id="cb86-2" title="2">mrks &lt;-<span class="st"> </span><span class="kw">t</span>(genos[<span class="dv">4</span><span class="op">:</span><span class="kw">ncol</span>(genos)])</a>
<a class="sourceLine" id="cb86-3" title="3">mrks &lt;-<span class="st"> </span>mrks <span class="op">+</span><span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb86-4" title="4">myGD &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">Taxa =</span> <span class="kw">colnames</span>(genos)[<span class="dv">4</span><span class="op">:</span><span class="kw">ncol</span>(genos)], mrks)</a>
<a class="sourceLine" id="cb86-5" title="5"><span class="kw">colnames</span>(myGD)[<span class="dv">2</span><span class="op">:</span><span class="kw">ncol</span>(myGD)] &lt;-<span class="st"> </span><span class="kw">as.character</span>(genos<span class="op">$</span>rs)</a>
<a class="sourceLine" id="cb86-6" title="6"></a>
<a class="sourceLine" id="cb86-7" title="7"><span class="kw">colnames</span>(phenos)[<span class="dv">1</span>] &lt;-<span class="st"> &quot;Taxa&quot;</span></a></code></pre></div>
</div>
<div id="gwas-with-gapit" class="section level4">
<h4>GWAS with GAPIT</h4>
<p>The best thing about GAPIT is that it offers several algorithms for GWAS. While these all essentially do the same thing (solves a LMM), there are certainly differneces in performance between algorithms. Below is a brief summary of the methods:</p>
<ul>
<li>LMM: This is the base LMM that is run with GWAS.</li>
<li>P3D: The P3D method estimates variance components once using a model without SNP effects and solutions are used for each SNP</li>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2931336/">Compressed Mixed linear model (CMLM)</a>: Rather thank calculating the kinship between individuals, CMLM clusters the accessions into smaller groups and estimates the relationships between groups.</li>
<li><a href="https://bmcbiol.biomedcentral.com/articles/10.1186/s12915-014-0073-5">Enriched CMLM</a>: Provides additional clustering approaches</li>
<li><a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0107684">Super GWAS</a>: Uses a subset of markers to define the relationship</li>
<li>Multiple Loci Mixed linear Model (MLMM): Uses step-wise linear regression to add SNP to the model</li>
<li>FarmCPU</li>
</ul>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb87-1" title="1">myGAPIT &lt;-<span class="st"> </span><span class="kw">GAPIT</span>(<span class="dt">Y =</span> phenos, <span class="dt">GD =</span> myGD, <span class="dt">GM =</span> myGM, <span class="dt">PCA.total =</span> <span class="dv">0</span>, <span class="dt">SNP.MAF =</span> <span class="fl">0.05</span>) <span class="co">#The default model uses CMLM</span></a>
<a class="sourceLine" id="cb87-2" title="2"></a>
<a class="sourceLine" id="cb87-3" title="3">myGAPIT2 &lt;-<span class="st"> </span><span class="kw">GAPIT</span>(<span class="dt">Y =</span> phenos, <span class="dt">GD =</span> myGD, <span class="dt">GM =</span> myGM,</a>
<a class="sourceLine" id="cb87-4" title="4">                  <span class="dt">kinship.cluster =</span> <span class="st">&quot;ward.D&quot;</span>, <span class="co">#how to cluster accessions</span></a>
<a class="sourceLine" id="cb87-5" title="5">                  <span class="dt">kinship.group =</span> <span class="st">&quot;Max&quot;</span>, <span class="co">#how to calculate kinship between clusters</span></a>
<a class="sourceLine" id="cb87-6" title="6">                  <span class="dt">group.from =</span> <span class="dv">1</span>, <span class="co">#Min number of groups</span></a>
<a class="sourceLine" id="cb87-7" title="7">                  <span class="dt">group.to =</span> <span class="kw">nrow</span>(phenos), <span class="co">#Max number of groups</span></a>
<a class="sourceLine" id="cb87-8" title="8">                  <span class="dt">group.by =</span> <span class="dv">10</span>)</a></code></pre></div>
</div>
</div>
</div>
<div id="determining-an-appropriate-p-value-threshold" class="section level2">
<h2>Determining an appropriate p-value threshold</h2>
<p>Here, we will use the Meff method from Li and Ji (2005) to calculate the appropriate p-value threshold. Since we know markers on separate chromosomes should be in LE, we will subset the data by chromosome.</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb88-1" title="1">Li.Ji &lt;-<span class="st"> </span><span class="cf">function</span>(CORMAT){</a>
<a class="sourceLine" id="cb88-2" title="2">  abs.evs &lt;-<span class="st"> </span><span class="kw">abs</span>(<span class="kw">eigen</span>(CORMAT, <span class="dt">only.values=</span><span class="ot">TRUE</span>)<span class="op">$</span>values) <span class="co">#Get eigenvalues for correlation matrix</span></a>
<a class="sourceLine" id="cb88-3" title="3">  Meff &lt;-<span class="st"> </span><span class="kw">sum</span>( <span class="kw">ifelse</span>(abs.evs <span class="op">&gt;=</span><span class="st"> </span><span class="dv">1</span>, <span class="dv">1</span>, (abs.evs <span class="op">-</span><span class="st"> </span><span class="kw">floor</span>(abs.evs))))</a>
<a class="sourceLine" id="cb88-4" title="4">  <span class="kw">return</span>(Meff)</a>
<a class="sourceLine" id="cb88-5" title="5">}</a>
<a class="sourceLine" id="cb88-6" title="6"></a>
<a class="sourceLine" id="cb88-7" title="7">nChr &lt;-<span class="st"> </span><span class="kw">length</span>(<span class="kw">unique</span>(genos<span class="op">$</span>chr))</a>
<a class="sourceLine" id="cb88-8" title="8"></a>
<a class="sourceLine" id="cb88-9" title="9">Meffs &lt;-<span class="st"> </span><span class="ot">NULL</span></a>
<a class="sourceLine" id="cb88-10" title="10"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>nChr){</a>
<a class="sourceLine" id="cb88-11" title="11">  tmp &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(genos[genos<span class="op">$</span>chr <span class="op">%in%</span><span class="st"> </span>i ,][<span class="dv">4</span><span class="op">:</span><span class="kw">ncol</span>(genos)])</a>
<a class="sourceLine" id="cb88-12" title="12">  </a>
<a class="sourceLine" id="cb88-13" title="13">  tmp &lt;-<span class="st"> </span><span class="kw">cor</span>(<span class="kw">t</span>(tmp))</a>
<a class="sourceLine" id="cb88-14" title="14">  </a>
<a class="sourceLine" id="cb88-15" title="15">  Meffs &lt;-<span class="st"> </span><span class="kw">c</span>(Meffs, <span class="kw">Li.Ji</span>(tmp))</a>
<a class="sourceLine" id="cb88-16" title="16">}</a>
<a class="sourceLine" id="cb88-17" title="17"></a>
<a class="sourceLine" id="cb88-18" title="18">Meffs &lt;-<span class="st"> </span><span class="kw">sum</span>(Meffs)</a>
<a class="sourceLine" id="cb88-19" title="19"><span class="kw">print</span>(Meffs) <span class="co">#~3900</span></a>
<a class="sourceLine" id="cb88-20" title="20"></a>
<a class="sourceLine" id="cb88-21" title="21">p.thresh &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span><span class="fl">0.05</span>)<span class="op">^</span>(<span class="dv">1</span><span class="op">/</span>Meffs) <span class="co">#1.315216e-05</span></a></code></pre></div>
</div>
</div>


    </div>
    <div class="col-md-2">
            <div id="toc">
      	<button type="button" class="close">×</button>
        <ul>
        <li><a href="#imputed-rice-data">Imputed rice data</a></li>
        <li><a href="#gwas">GWAS</a></li>
        <li><a href="#determining-an-appropriate-p-value-threshold">Determining an appropriate p-value threshold</a></li>
        </ul>
      </div>
          </div>
  </div>
</div>

<script>
$(document).ready(function () {

  // add bootstrap table styles to pandoc tables
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');

});
</script>

<script>
$(document).ready(function () {
 	 	$('#content img')
 	  .addClass("image-thumb");
      $('#content img')
 	  .addClass("image-lb");
  $('#content').magnificPopup({
	      type:'image',
	      closeOnContentClick: false,
	      closeBtnInside: false,
	      delegate: 'img',
	      gallery: {enabled: false },
	      image: {
	        verticalFit: true,
          titleSrc: 'alt'
	      }
 	    });
 	});
</script>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
